src/circuit/operator/plus.rs

```diff
@@ -0,0 +1,290 @@
+//! Binary plus operator.
+
+use crate::{
+    algebra::{AddAssignByRef, AddByRef},
+    circuit::{
+        operator_traits::{BinaryOperator, Operator, UnaryOperator},
+        Circuit, OwnershipPreference, Stream,
+    },
+};
+use std::{borrow::Cow, marker::PhantomData, ops::Add};
+
+impl<P, D> Stream<Circuit<P>, D>
+where
+    P: Clone + 'static,
+    D: Add<Output = D> + AddByRef + AddAssignByRef + Clone + 'static,
+{
+    /// Apply the [`Plus`] operator to `self` and `other`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use dbsp::circuit::{
+    /// #     operator::Generator,
+    /// #     Root,
+    /// # };
+    /// let root = Root::build(move |circuit| {
+    ///     // Stream of non-negative values: 0, 1, 2, ...
+    ///     let mut n = 0;
+    ///     let source1 = circuit.add_source(Generator::new(move || {let res = n; n += 1; res}));
+    ///     // Stream of non-positive values: 0, -1, -2, ...
+    ///     let mut n = 0;
+    ///     let source2 = circuit.add_source(Generator::new(move || {let res = n; n -= 1; res}));
+    ///     // Compute pairwise sums of values in the stream; the output stream will contain zeros.
+    ///     source1.plus(&source2).inspect(|n| assert_eq!(*n, 0));
+    /// })
+    /// .unwrap();
+    ///
+    /// # for _ in 0..5 {
+    /// #     root.step().unwrap();
+    /// # }
+    /// ```
+    pub fn plus(&self, other: &Stream<Circuit<P>, D>) -> Stream<Circuit<P>, D> {
+        if self.origin_node_id() == other.origin_node_id() {
```

Comment:
you are embedding a little optimizer here, I think this should really be the job of the compiler, which may have a cost model.

Reply:
I agree. This is why I only implemented a really simplistic thing that's necessary to efficiently implement methods like `integrate`.  Anything beyond that is up to the comiler.